### PR TITLE
Add `-noReconnectAfter` option

### PR DESCRIPTION
**Pull request by Vlatombe**
This option allows to abort retrying after a given timeout expressed in simple form (e.g. `1d`, `2h`, `10m`, `60s`, ...)
 
~~Downstream work to include this option in the docker image for agents: https://github.com/jenkinsci/docker-agent/pull/802~~

### Testing done

Interactive testing

* Inbound TCP
  * On initial connection ✅
  * On disconnect ✅

* Websocket
  * Initial connection ❌: `-noReconnect` not implemented
  * On disconnect ✅

```[tasklist]
### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Comment by jglick** [comment]
Does this need to be an option, or can we just hard-code a timeout of (say) one hour? I guess that could break some use case involving a static agent connected to a controller which is down for maintenance over a long weekend?

**Comment by Vlatombe** [comment]
I could give the option a sensible default value, but I think one hour may be too short, considering some maintenance followed by a startup sequence on slow hardware for the controller. Maybe one day?

**Comment by jglick** [comment]
Not sure. I guess making the timeout opt-in makes sense in that it can be set by all cloud plugins, as a complement for example to https://github.com/jenkinsci/kubernetes-plugin/pull/1543; and anyone using static agents can decide for themselves if it is helpful.

File: src/main/java/hudson/remoting/DurationSecondsOptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import java.time.Duration;
+import org.kohsuke.args4j.CmdLineException;
+import org.kohsuke.args4j.CmdLineParser;
+import org.kohsuke.args4j.OptionDef;
+import org.kohsuke.args4j.spi.OptionHandler;
+import org.kohsuke.args4j.spi.Parameters;
+import org.kohsuke.args4j.spi.Setter;
+
+/**
+ * Parses a positive number of seconds into a {@link Duration}.
```
Comment (jglick) [inline src/main/java/hudson/remoting/DurationSecondsOptionHandler.java:12]: If we are bothering to introduce a custom option handler, why not make it friendlier and accept `3h`, `1d`, etc.? Is there some official format for concise duration strings?

File: src/main/java/hudson/remoting/Engine.java [hunk 2]
```diff
@@ -740,9 +750,14 @@ public void closeRead() throws IOException {
                 if (noReconnect) {
                     return;
                 }
+                firstAttempt = Instant.now();
```
Comment (jglick) [inline src/main/java/hudson/remoting/Engine.java:753]: (WS reconnect)

File: src/main/java/hudson/remoting/Engine.java [hunk 3]
```diff
@@ -795,14 +810,18 @@ private void innerRun(IOHub hub, SSLContext context, ExecutorService service) {
 
         try {
             boolean first = true;
+            firstAttempt = Instant.now();
```
Comment (jglick) [inline src/main/java/hudson/remoting/Engine.java:813]: TCP initial connection?

File: src/main/java/hudson/remoting/Engine.java [hunk 4]
```diff
@@ -915,7 +934,7 @@ private void innerRun(IOHub hub, SSLContext context, ExecutorService service) {
                 }
                 if(noReconnect)
                     return; // exit
-
+                firstAttempt = Instant.now();
```
Comment (jglick) [inline src/main/java/hudson/remoting/Engine.java:937]: TCP reconnect?

File: src/main/java/hudson/remoting/Launcher.java [hunk 5]
```diff
@@ -401,6 +406,9 @@ public static void main(String... args) throws IOException, InterruptedException
                 System.err.println(
                         "WARNING: Providing the secret and agent name as positional arguments is deprecated; use \"-secret\" and \"-name\" instead.");
             }
+            if (launcher.noReconnect && launcher.noReconnectAfter != null) {
+                throw new CmdLineException(null, "-noReconnect and -noReconnectAfter are mutually exclusive");
+            }
```
Comment (jglick) [inline src/main/java/hudson/remoting/Launcher.java:411]: rather use `forbids` in the annotation

File: src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java [hunk 6]
```diff
@@ -401,8 +407,13 @@ public void waitForReady() throws InterruptedException {
         String oldName = t.getName();
         try {
             int retries = 0;
+            Instant firstAttempt = Instant.now();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java:410]: (TCP discovery via HTTP)

**Review by jglick — APPROVED** [review]
Looks good so far.

**Comment by jglick** [comment]
(complements #676 I think)

File: src/main/java/hudson/remoting/DurationSecondsOptionHandler.java (hunk 1 above)
Comment (Vlatombe) [reply src/main/java/hudson/remoting/DurationSecondsOptionHandler.java:12]: spring boot has a utility for this we could copy
https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/convert/DurationStyle.java

File: src/main/java/org/jenkinsci/remoting/util/DurationStyle.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+                if (candidate.chronoUnit == chronoUnit) {
+                    return candidate;
+                }
+            }
+            throw new IllegalArgumentException("Unknown unit " + chronoUnit);
+        }
+
+        public static Unit fromSuffix(String suffix) {
+            for (Unit candidate : values()) {
+                if (candidate.suffix.equals(suffix.toLowerCase())) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/util/DurationStyle.java:254]: need to use `Locale.ROOT`

File: src/main/java/org/jenkinsci/remoting/DurationOptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+import org.kohsuke.args4j.CmdLineParser;
+import org.kohsuke.args4j.OptionDef;
+import org.kohsuke.args4j.spi.OptionHandler;
+import org.kohsuke.args4j.spi.Parameters;
+import org.kohsuke.args4j.spi.Setter;
+
+/**
+ * Parses a string like 1s, 2m, 3h, 4d into a {@link Duration}.
+ */
+public class DurationOptionHandler extends OptionHandler<Duration> {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/DurationOptionHandler.java:39]: May want to mark new util classes `NoExternalUse`.

File: src/main/java/org/jenkinsci/remoting/util/DurationFormatter.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+
+/**
+ * Formats a {@link Duration} into a human-readable string.
+ */
+public final class DurationFormatter {
+    private DurationFormatter(){}
+
+    public static String format(Duration d) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/util/DurationFormatter.java:35]: BTW there is also https://github.com/jenkinsci/jenkins/blob/f8e381fb44d59432d654bc64eb314af7d23f79e9/core/src/main/java/hudson/Util.java#L736-L745

File: src/main/java/org/jenkinsci/remoting/util/DurationStyle.java [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+        /**
+         * Nanoseconds.
+         */
+        NANOS(ChronoUnit.NANOS, "ns", Duration::toNanos),
+
+        /**
+         * Microseconds.
+         */
+        MICROS(ChronoUnit.MICROS, "us", (duration) -> duration.toNanos() / 1000L),
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/util/DurationStyle.java:189]: Should really be 
```suggestion
        MICROS(ChronoUnit.MICROS, "µs", (duration) -> duration.toNanos() / 1000L),
```
! Oh well.

**Review by jglick — APPROVED** [review]

**Comment by jglick** [comment]
Is there a corresponding PR to `kubernetes-plugin` to activate this mode? I guess there is no simple way to consume https://github.com/jenkinsci/docker-agent/pull/809 prior to merge & release?

**Comment by Vlatombe** [comment]
Unfortunately no, there isn't.

**Pull request merged**